### ISSUE TITLE
ParserModel doesn't capture control flow

### ISSUE THREAD

**Issue by jnfoster** [p4lang/p4-spec#239]
According to the ParserModel:

- `transition e` is modeled as
```
ParserModel.currentState = eval(e)
```
- `verify(b, err)` is modeled as
 
```
    if (b == false) {
        ParserModel.parserError = err;
        ParserModel.currentState = reject;
    }
```
- and statements are evaluated in sequential order.

Putting these together, it follows that` verify(false, Oops); transition accept` is modeled as a program equivalent to:
```
ParserModel.parserError = Oops;
ParserModel.currentState = reject;
ParserModel.currentState = accept;
```
which isn't at all what we meant!

Unless we want to CPS convert everything or something, it won't be easy to model the intended semantics. I wonder if we should give up trying to have a complete abstract machine for parsers and just give pseudo code for some of their components.

**Comment by jafingerhut** [comment]
From this statement in section 11.7 "verify"

"If the first argument is false, it causes an immediate transition to the reject state, which causes immediate parsing termination; at the same time, the parserError associated with the parser is set to the value of the second argument."

My understanding is that the 'transition accept' statement in your example would never be executed, because the verify "causes immediate parsing termination".

**Comment by jnfoster** [comment]
The problem is that our pseudo-code is we are not modeling control-flow property. This is easy to see: the target of the translation (P4 `parsers`) uses exceptions but the target of the translation (imperative pseudocode) does not have them. 

So our translation is just wrong.

It is possible to "translate away" exceptions, but you generally need to do some kind of global transformation, like conversion to continuation passing style (CPS). 

I agree that it's obvious what we mean here, but it's embarrassing (for me) to get this sort of detail wrong. These ideas are elementary in the programming language literature, and if we'd done our job correctly, someone who wanted to build a reference implementation would be able to just translate our pseudocode into something executable. That's not the case with what we have now. Following the text carefully, you can end up with "transitions to reject" being overwritten, since they are modeled by simply modifying an internal state variable (and when there are two modifications to the same state variable the second overwrites the effects of the first, which is different than how exceptions behave.)

**Comment by mbudiu-vmw** [comment]
There are no exceptions in the language, period.
I don't know what you are talking about.
The semantics of parsers is very simple, transition is just a goto.

**Comment by jnfoster** [comment]
You're right -- I was thinking of `valid` as throwing an exception which is not literally true. It's a non-local control flow.

However, the crux of the issue remains. The concrete example I wrote down above is wrong in ParserModel as we describe it. The target language of our translation uses assignment, not `goto`s, so we get the wrong result for:

```
verify(false, Oops);
transition accept
```

**Comment by jafingerhut** [comment]
@jnfoster - If the implementation of verify contains an 'exit' statement that does a complete exit from the parser control flow, that sounds like correct behavior to me (even if 'exit' is not something a P4 program can contain explicitly in parser code).

**Comment by jnfoster** [comment]
Yep. We can add exceptions, gotos, or `halt` (maybe to avoid confusion with P4's `exit`) and fix-up the translation.

**Comment by mbudiu-vmw** [comment]
The "state" is the program counter. This program uses assignments to the program counter.
There is no halt or exit needed.

**Comment by jnfoster** [comment]
@mbudiu-vmw Can you show me what 
```
verify(false, Oops);
transition accept;
```
looks like in the ParserModel pseudo-code? I have a counter-example. (I don't see how the PC helps us since we say that multiple statements are executed sequentially.)

**Comment by jnfoster** [comment]
Resolved: to replace assignment to currentState with a goto.

**Comment by jafingerhut** [comment]
Opened this PR with proposed change: https://github.com/p4lang/p4-spec/pull/248

**Comment by jnfoster** [comment]
Fixed by #248
